Userland/Services/DeviceMapper/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,241 @@
+    }
+    auto removed_anything = family.registered_nodes().remove_all_matching([minor_number](auto& device) { return device.minor_number() == minor_number; });
+    if (!removed_anything) {
+        // FIXME: Handle cases where we can't remove a device node.
+        // This could happen when the DeviceMapper program was restarted
+        // so the previous state was not preserved and a device was removed.
+        return Error::from_errno(ENODEV);
+    }
+    return {};
+}
+
+static ErrorOr<void> create_kcov_device_node()
+{
+    mode_t old_mask = umask(0);
+    TRY(Core::System::create_char_device("/dev/kcov"sv, 0666, 30, 0));
+    umask(old_mask);
+    return {};
+}
+
+ErrorOr<void> DeviceEventLoop::read_one_or_eof(DeviceEvent& event)
+{
+    if (m_devctl_file->read_until_filled({ bit_cast<u8*>(&event), sizeof(DeviceEvent) }).is_error()) {
+        // Bad! Kernel and SystemServer apparently disagree on the record size,
+        // which means that previous data is likely to be invalid.
+        return Error::from_string_view("File ended after incomplete record? /dev/devctl seems broken!"sv);
+    }
+    return {};
+}
+
+ErrorOr<void> DeviceEventLoop::drain_events_from_devctl()
+{
+    for (;;) {
+        DeviceEvent event;
+        TRY(read_one_or_eof(event));
+        // NOTE: Ignore any event related to /dev/devctl device node - normally
+        // it should never disappear from the system and we already use it in this
+        // code.
+        if (event.major_number == 2 && event.minor_number == 10 && !event.is_block_device)
+            continue;
+
+        if (event.state == DeviceEvent::State::Inserted) {
+            // NOTE: We have a special function for the KCOV device, because we don't
+            // want to create a new MinorNumberAllocationType (e.g. SingleInstance).
+            // Instead, just blindly create that device node and assume we will never
+            // have to worry about it, so we don't need to register that!
+            if (event.major_number == 30 && event.minor_number == 0 && !event.is_block_device) {
+                TRY(create_kcov_device_node());
+                continue;
+            }
+            VERIFY(event.is_block_device == 1 || event.is_block_device == 0);
+            TRY(register_new_device(event.is_block_device, event.major_number, event.minor_number));
+        } else if (event.state == DeviceEvent::State::Removed) {
+            if (event.major_number == 30 && event.minor_number == 0 && !event.is_block_device) {
+                dbgln("DeviceMapper: unregistering device failed: kcov tried to de-register itself!?");
+                continue;
+            }
+            if (auto error_or_void = unregister_device(event.is_block_device, event.major_number, event.minor_number); error_or_void.is_error())
+                dbgln("DeviceMapper: unregistering device failed: {}", error_or_void.error());
+        } else {
+            dbgln("DeviceMapper: Unhandled device event!");
```

Comment:
Suggestion: add `event.state` in the `dbgln` format so it's immediately clear which state triggered this.